### PR TITLE
Include const generic arguments in metadata

### LINKED ISSUE #61624: ICE using a type alias to a const generic struct across crates
In a folder `a`, I have a blank library crate with the following `src/lib.rs`:
```rust
#![feature(const_generics)]

pub struct Struct<const N: usize>([u8; N]);

pub type Alias = Struct<2>;
```
In a folder `b`, I have a blank binary crate with the follow `src/main.rs`:
```rust
// note: the error seems to be the same regardless of whether the following line is commented or not
//#![feature(const_generics)]

struct Container(a::Alias);

fn main() {}
```
`b`'s Cargo.toml uses crate `a` with the following line: 
```
[dependencies]
a = { path = "../a", version = "" }
```
Running `cargo check` or `cargo build` in `b` will produce the following ICE:
```
    Checking b v0.1.0 (/home/carado/tmp/rust7/b)
error: internal compiler error: src/librustc_metadata/decoder.rs:484: entry: id not found: DefIndex(18) in crate a with number 15

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:638:9
stack backtrace:
   0:     0x7f4371ffb19b - backtrace::backtrace::libunwind::trace::h034f17449c8f06d2
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.29/src/backtrace/libunwind.rs:88
   1:     0x7f4371ffb19b - backtrace::backtrace::trace_unsynchronized::h88fceba984274b2d
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.29/src/backtrace/mod.rs:66
   2:     0x7f4371ffb19b - std::sys_common::backtrace::_print::h113a7251db1bb0bb
                               at src/libstd/sys_common/backtrace.rs:47
   3:     0x7f4371ffb19b - std::sys_common::backtrace::print::he1709f3e64aa1331
                               at src/libstd/sys_common/backtrace.rs:36
   4:     0x7f4371ffb19b - std::panicking::default_hook::{{closure}}::h4c7c19a95acf509e
                               at src/libstd/panicking.rs:198
   5:     0x7f4371ffae9c - std::panicking::default_hook::haf2a960f108a3f30
                               at src/libstd/panicking.rs:212
   6:     0x7f436fd54dd1 - rustc::util::common::panic_hook::h96cf50c713bc84f6
   7:     0x7f4371ffb9f9 - std::panicking::rust_panic_with_hook::h89241749267406d1
                               at src/libstd/panicking.rs:479
   8:     0x7f436e9a652d - std::panicking::begin_panic::h23f047602abbec1f
   9:     0x7f436e9c661f - rustc_errors::Handler::bug::h1846692de954732d
  10:     0x7f436f9ceac3 - rustc::util::bug::opt_span_bug_fmt::{{closure}}::h4eeb8977ea9f490d
  11:     0x7f436f9ce09a - rustc::ty::context::tls::with_opt::{{closure}}::h4f8aad466ccd8399
  12:     0x7f436f9cdfb5 - rustc::ty::context::tls::with_context_opt::h93eb33abb81e1499
  13:     0x7f436f9ce047 - rustc::ty::context::tls::with_opt::h84f79a4206f98907
  14:     0x7f436f9ce9d8 - rustc::util::bug::opt_span_bug_fmt::hcae1e85db26d61d1
  15:     0x7f436f9ce942 - rustc::util::bug::bug_fmt::h0d9f8eae974cd447
  16:     0x7f437048e53a - rustc_metadata::decoder::<impl rustc_metadata::cstore::CrateMetadata>::entry::hd6886155c7e8ad94
  17:     0x7f437048f82c - rustc_metadata::decoder::<impl rustc_metadata::cstore::CrateMetadata>::get_type::he89c8116f2a9d946
  18:     0x7f4370499601 - rustc_metadata::cstore_impl::provide_extern::type_of::h81daf21c535e688b
  19:     0x7f436fce71cd - rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::type_of>::compute::h6309e2152a5e3c65
  20:     0x7f436f8b5517 - rustc::dep_graph::graph::DepGraph::with_task_impl::had7a0c15c5f1905b
  21:     0x7f436fbb7456 - rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query::hb3169834330de254
  22:     0x7f436f7cea4f - rustc::ty::instance::Instance::resolve::h72798ee26e58abe1
  23:     0x7f436fd411bb - <rustc::traits::project::AssocTypeNormalizer as rustc::ty::fold::TypeFolder>::fold_const::hd5c5bcc5cc84d8c5
  24:     0x7f436f956ffa - <smallvec::SmallVec<A> as core::iter::traits::collect::FromIterator<<A as smallvec::Array>::Item>>::from_iter::h56ca20c895cf1fe3
  25:     0x7f436fa90fec - rustc::ty::fold::TypeFoldable::fold_with::h9620990ca94be452
  26:     0x7f436faacdcd - rustc::ty::structural_impls::<impl rustc::ty::fold::TypeFoldable for &rustc::ty::TyS>::super_fold_with::h42aca40e5d6f720c
  27:     0x7f436fd40b65 - <rustc::traits::project::AssocTypeNormalizer as rustc::ty::fold::TypeFolder>::fold_ty::h0000082e2319606a
  28:     0x7f436cd1ed64 - rustc::traits::project::normalize::hbd56f85fc248fb4d
  29:     0x7f436cb40590 - rustc_typeck::check::FnCtxt::normalize_associated_types_in::h8d3d18278d4fdeda
  30:     0x7f436cd664b0 - <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::fold::h7a0fbf29f0ae4dd7
  31:     0x7f436ccb1f67 - rustc::ty::context::GlobalCtxt::enter_local::h05475195df7a61f1
  32:     0x7f436cd92e9b - rustc_typeck::check::wfcheck::check_item_well_formed::hdb36490216ee054e
  33:     0x7f436cb6165e - rustc::ty::query::__query_compute::check_item_well_formed::he41baf5c84b70167
  34:     0x7f436cc2b7bd - rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::check_item_well_formed>::compute::h7c840c44c057109d
  35:     0x7f436cb966fb - rustc::dep_graph::graph::DepGraph::with_task_impl::h9a2f7b5bbc29bc70
  36:     0x7f436cc876f0 - rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query::ha95df13349dc610e
  37:     0x7f436cd96967 - <rustc_typeck::check::wfcheck::CheckTypeWellFormedVisitor as rustc::hir::itemlikevisit::ParItemLikeVisitor>::visit_item::ha1c79ec81b0ae69c
  38:     0x7f437200c8ea - __rust_maybe_catch_panic
                               at src/libpanic_unwind/lib.rs:85
  39:     0x7f436cba2519 - rustc_data_structures::sync::par_for_each_in::he5330af9a7aa0673
  40:     0x7f437200c8ea - __rust_maybe_catch_panic
                               at src/libpanic_unwind/lib.rs:85
  41:     0x7f436cd59313 - rustc::hir::Crate::par_visit_all_item_likes::h7e35cc6980d7d055
  42:     0x7f436cb64fcd - rustc::util::common::time::h7201fa0034b9369a
  43:     0x7f436cd78a28 - rustc_typeck::check_crate::h2869f0da763831fb
  44:     0x7f43719ceb45 - rustc_interface::passes::analysis::h3924fc0f51a4916f
  45:     0x7f437235b576 - rustc::ty::query::__query_compute::analysis::hecebb52178cb72d8
  46:     0x7f43722c5e02 - rustc::dep_graph::graph::DepGraph::with_task_impl::h13f978acfb884bfb
  47:     0x7f43722bfe0d - rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query::h41a19a7094d84ad5
  48:     0x7f43722cfdc6 - rustc::ty::context::tls::enter_global::hde76617de84f513e
  49:     0x7f43722ee377 - rustc_interface::passes::BoxedGlobalCtxt::access::{{closure}}::h7e90e4645ca62211
  50:     0x7f4371a2c4d5 - rustc_interface::passes::create_global_ctxt::{{closure}}::hd1ffcd24d8253592
  51:     0x7f43722ef6ba - rustc_interface::interface::run_compiler_in_existing_thread_pool::hbb3327ac4471db7f
  52:     0x7f4372361346 - std::thread::local::LocalKey<T>::with::h889b3604823f7d6d
  53:     0x7f4372304e65 - scoped_tls::ScopedKey<T>::set::h04ec88caee9f9ad7
  54:     0x7f437232cea4 - syntax::with_globals::h380e3542e6008820
  55:     0x7f437234bd9d - std::sys_common::backtrace::__rust_begin_short_backtrace::h5f9f48a6f860c8d7
  56:     0x7f437200c8ea - __rust_maybe_catch_panic
                               at src/libpanic_unwind/lib.rs:85
  57:     0x7f43722d7f29 - core::ops::function::FnOnce::call_once{{vtable.shim}}::hc55a01470f319106
  58:     0x7f4371fdeebf - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::h3f73aebf71bfd2d8
                               at /rustc/5eeb567a27eba18420a620ca7d0c007e29d8bc0c/src/liballoc/boxed.rs:746
  59:     0x7f437200b5c0 - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::hf5740d907353c675
                               at /rustc/5eeb567a27eba18420a620ca7d0c007e29d8bc0c/src/liballoc/boxed.rs:746
  60:     0x7f437200b5c0 - std::sys_common::thread::start_thread::h9d627f833866d0b2
                               at src/libstd/sys_common/thread.rs:13
  61:     0x7f437200b5c0 - std::sys::unix::thread::Thread::new::thread_start::h6883c5e892c614cf
                               at src/libstd/sys/unix/thread.rs:79
  62:     0x7f4371f42fa8 - start_thread
                               at /builddir/glibc-2.29/nptl/pthread_create.c:486
  63:     0x7f4371e64a1f - __GI___clone
  64:                0x0 - <unknown>
query stack during panic:
#0 [type_of] processing `a::Alias::{{constant}}#0`
#1 [check_item_well_formed] processing `Container`
#2 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to previous error


note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.37.0-nightly (5eeb567a2 2019-06-06) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental --crate-type bin

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `b`.

To learn more, run the command again with --verbose.
```

### LINKED ISSUE #64707: ICE with const_generic and integration test
Searched a bit, but did not find the same issue. When using const_generics and trying to get an integration test (or benchmark) to run, the compiler ICEs.

I tried this code:
src/lib.rs:
```
#![feature(const_generics)]

pub struct Foo<const D: usize> {
    pub coords: [u8; D],
}

pub fn foo(_: &[Foo<{ 1 }>]) {}
```
tests/foo.rs:
```
extern crate rust_ice;

#[test]
fn test_foo() {
    rust_ice::foo(&vec![]);
}
```
Cargo.toml:
```
[package]
name = "rust_ice"
version = "0.1.0"
edition = "2018"
```
Commandline:
`CARGO_INCREMENTAL=0 cargo test`

I expected to see this happen: It should build and the test should run without any problems.

Instead, this happened: The Compiler crashes with the following error:
`error: internal compiler error: src/librustc_metadata/decoder.rs:474: entry: id not found: DefIndex(17) in crate rust_ice with number 15`

## Meta

`rustc --version --verbose`:
rustc 1.39.0-nightly (1dd188489 2019-09-22)
binary: rustc
commit-hash: 1dd1884891636d0eb51157d137230076bcf20627
commit-date: 2019-09-22
host: x86_64-unknown-linux-gnu
release: 1.39.0-nightly
LLVM version: 9.0

Backtrace:
```
error: internal compiler error: src/librustc_metadata/decoder.rs:474: entry: id not found: DefIndex(17) in crate rust_ice with number 15

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:643:9
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.37/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.37/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:76
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:60
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1030
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1412
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:64
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:196
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:210
  10: rustc_driver::report_ice
  11: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:477
  12: std::panicking::begin_panic
  13: rustc_errors::Handler::bug
  14: rustc::util::bug::opt_span_bug_fmt::{{closure}}
  15: rustc::ty::context::tls::with_opt::{{closure}}
  16: rustc::ty::context::tls::with_context_opt
  17: rustc::ty::context::tls::with_opt
  18: rustc::util::bug::opt_span_bug_fmt
  19: rustc::util::bug::bug_fmt
  20: rustc_metadata::cstore_impl::provide_extern::type_of
  21: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::type_of>::compute
  22: rustc::dep_graph::graph::DepGraph::with_task_impl
  23: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  24: rustc::ty::instance::Instance::resolve
  25: rustc::ty::sty::Const::eval
  26: <smallvec::SmallVec<A> as core::iter::traits::collect::FromIterator<<A as smallvec::Array>::Item>>::from_iter
  27: rustc::ty::fold::TypeFoldable::fold_with
  28: <rustc::traits::project::AssocTypeNormalizer as rustc::ty::fold::TypeFolder>::fold_ty
  29: <rustc::traits::project::AssocTypeNormalizer as rustc::ty::fold::TypeFolder>::fold_ty
  30: <rustc::traits::project::AssocTypeNormalizer as rustc::ty::fold::TypeFolder>::fold_ty
  31: <smallvec::SmallVec<A> as core::iter::traits::collect::FromIterator<<A as smallvec::Array>::Item>>::from_iter
  32: rustc::ty::fold::TypeFoldable::fold_with
  33: rustc::traits::project::normalize
  34: rustc_typeck::check::callee::<impl rustc_typeck::check::FnCtxt>::confirm_builtin_call
  35: rustc_typeck::check::callee::<impl rustc_typeck::check::FnCtxt>::check_call
  36: rustc_typeck::check::expr::<impl rustc_typeck::check::FnCtxt>::check_expr_kind
  37: rustc_typeck::check::expr::<impl rustc_typeck::check::FnCtxt>::check_expr_with_expectation_and_needs
  38: rustc_typeck::check::FnCtxt::check_stmt
  39: rustc_typeck::check::FnCtxt::check_block_with_expected
  40: rustc_typeck::check::expr::<impl rustc_typeck::check::FnCtxt>::check_expr_kind
  41: rustc_typeck::check::expr::<impl rustc_typeck::check::FnCtxt>::check_expr_with_expectation_and_needs
  42: rustc_typeck::check::expr::<impl rustc_typeck::check::FnCtxt>::check_return_expr
  43: rustc_typeck::check::check_fn
  44: rustc::ty::context::GlobalCtxt::enter_local
  45: rustc_typeck::check::typeck_tables_of
  46: rustc::ty::query::__query_compute::typeck_tables_of
  47: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::typeck_tables_of>::compute
  48: rustc::dep_graph::graph::DepGraph::with_task_impl
  49: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  50: rustc::ty::<impl rustc::ty::context::TyCtxt>::par_body_owners
  51: rustc_typeck::check::typeck_item_bodies
  52: rustc::ty::query::__query_compute::typeck_item_bodies
  53: rustc::dep_graph::graph::DepGraph::with_task_impl
  54: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  55: rustc::util::common::time
  56: rustc_typeck::check_crate
  57: rustc_interface::passes::analysis
  58: rustc::ty::query::__query_compute::analysis
  59: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  60: rustc_interface::passes::BoxedGlobalCtxt::access::{{closure}}
  61: rustc_interface::passes::create_global_ctxt::{{closure}}
  62: rustc_interface::interface::run_compiler_in_existing_thread_pool
  63: std::thread::local::LocalKey<T>::with
  64: scoped_tls::ScopedKey<T>::set
  65: syntax::with_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

### LINKED ISSUE #64730: error: internal compiler error: src/librustc_metadata/decoder.rs:483: entry: id not found: DefIndex(86) in crate aljabar with number 15
Hey everybody :) I was trying to test matrix*vector in aljabar crate by
running "cargo test" on:
```


extern crate aljabar;
use aljabar::{matrix, vector};



#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn identity() {
        let pixel = vector![0,255,255,15];
        let kernel = matrix![
        [0,0,0],
        [0,1,0],
        [0,0,0]];
        assert_eq!(pixel*kernel, pixel);
    }
}


```

  and got this:
> 
> error: internal compiler error: src/librustc_metadata/decoder.rs:483: entry: id not found: DefIndex(86) in crate aljabar with number 15
> 
> thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:644:9
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
> error: aborting due to previous error
> 
> 
> note: the compiler unexpectedly panicked. this is a bug.
> 
> note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
> 
> note: rustc 1.38.0-nightly (60960a260 2019-08-12) running on x86_64-apple-darwin
> 
> note: compiler flags: -C debuginfo=2 -C incremental
> 
> note: some of the compiler flags provided by cargo are hidden
> 
> error: Could not compile `KernelFiltering`.
> 
> 

Not sure if it's a known issue or not but if there's anything else you need from me please let me know :)

### PR DESCRIPTION
**Pull request by BenLewis-Seequent**
Fixes #64707
Fixes #61624
Fixes #64730

r? @varkor

### CODE REVIEW

File: src/test/ui/const-generics/const-argument-cross-crate-mismatch.stderr [hunk 1]
```diff
@@ -0,0 +1,21 @@
+error[E0308]: mismatched types
+  --> $DIR/const-argument-cross-crate-mismatch.rs:6:41
+   |
+LL |     let _ = const_generic_lib::function(const_generic_lib::Struct([0u8, 1u8]));
+   |                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `3usize`, found `2usize`
+   |
+   = note: expected type `const_generic_lib::Struct<3usize>`
+              found type `const_generic_lib::Struct<_: usize>`
```
Comment (Centril) [inline src/test/ui/const-generics/const-argument-cross-crate-mismatch.stderr]: @varkor What's up with `<_: usize>` here?

Comment (varkor) [reply src/test/ui/const-generics/const-argument-cross-crate-mismatch.stderr]: Not sure, as it seems to know what the value is. We can open another issue for it.

File: src/librustc_metadata/encoder.rs [hunk 2]
```diff
@@ -1708,8 +1708,19 @@ impl Visitor<'tcx> for EncodeContext<'tcx> {
         NestedVisitorMap::OnlyBodies(&self.tcx.hir())
     }
     fn visit_expr(&mut self, ex: &'tcx hir::Expr) {
-        intravisit::walk_expr(self, ex);
-        self.encode_info_for_expr(ex);
+        if let hir::ExprKind::Repeat(ref element, _) = ex.kind {
+            // The constant in a repeat expression doesn't need to be included in the metadata as
```
Comment (varkor) [inline src/librustc_metadata/encoder.rs]: Does this cause an issue if it is included? I.e. is it worth special-casing `ExprKind::Repeat` specifically?

**Review by varkor — APPROVED** [review]
@skinny121: this looks great, thank you! I'm really happy to see all the work you've been putting into fixing these const generics issues! :)

I just have one question, but I'm happy with these changes.

File: src/test/ui/const-generics/const-argument-cross-crate-mismatch.stderr (hunk 1 above)
Comment (varkor) [reply src/test/ui/const-generics/const-argument-cross-crate-mismatch.stderr]: Opened https://github.com/rust-lang/rust/issues/65380.

File: src/librustc_metadata/encoder.rs (hunk 2 above)
Comment (skinny121) [reply src/librustc_metadata/encoder.rs]: Don't think so, the only downside I think would be the impact on the size of the metadata, although it's probably not a significant impact.

Comment (skinny121) [reply src/librustc_metadata/encoder.rs]: I did some comparisons on the stdlib crates using `-Z meta_stats` option. It seems that it is indeed pretty insignificant as the increase in total bytes by including `ExprKind::Repeat` consts was typically around 0.01% - 0.05% with a maximum increase of 0.6%. So I have removed the special casing of `ExprKind::Repeat`.

**Comment by varkor** [comment]
@skinny121: if you could squash your two commits (as the second just undoes something in the first), that'd be great!

**Comment by varkor** [comment]
Thank you!

@bors r+ rollup

**Pull request merged**